bpo-34270: added the ability to name asyncio tasks

agronholm opened the pull request:
The ability to name individual tasks helps developers debug complex
asyncio applications.

Co-authored-by: Antti Haapala <antti.haapala@anttipatterns.com>

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-34270](https://www.bugs.python.org/issue34270) -->
https://bugs.python.org/issue34270
<!-- /issue-number -->

agronholm committed:
bpo-34270: added the "name" attribute to asyncio tasks

The ability to name individual tasks helps developers debug complex
asyncio applications.

Co-authored-by: Antti Haapala <antti.haapala@anttipatterns.com>

zooba commented on Lib/asyncio/tasks.py:
```diff
@@ -126,6 +131,14 @@ def __del__(self):
+
+    @name.setter
+    def name(self, value):
+        self._name = str(value)
```
Is forcing this to a `str` really important enough to need a property? The name is basically only ever used for printing anyway, which will do the conversion when necessary.

agronholm replied:
This is consistent with the behavior of `threading.Thread`. Some code might also access the `name` property directly, expecting a `str`.

1st1 replied:
As I mentioned above I don't think we need a setter. A `Task.get_name()` function is enough for now.

agronholm replied:
You can set names of threads and processes too, why not tasks as well? Where's the harm in that? I can think of a few use cases for this, like displaying the status of a task as it progresses.
I could convert this to java-style setters and getters if you want, despite how much I despise them (not to mention the fact that `threading` also switched away from this behavior). But your call.

1st1 replied:
> I can think of a few use cases for this, like displaying the status of a task as it progresses.

Hm, interesting. OK, that's at least one valid use-case :)

> I could convert this to java-style setters and getters if you want, despite how much I despise them (not to mention the fact that threading also switched away from this behavior). But your call.

It's not about liking Java-style getters or not, it's about consistency.  If asyncio's APIs were using `@property` it'd be equally happy about that, but now we better follow the overall style and use `get_name()` and `set_name()`.

zooba reviewed (approved):
Wait for Yury, but it looks good to me. One question about the need to convert names to str on assignment, but not a big deal.

elprans commented on Lib/asyncio/tasks.py:
```diff
@@ -23,6 +24,9 @@
 from .coroutines import coroutine
 
+# Helper to generate new task names
+_counter = itertools.count(1).__next__
```
Using `itertools.count()` seems like a bit of an overkill to me.   I'd simply do an explicit increment.  Also, `_task_name_counter` would be a clearer name.

1st1 replied:
Indeed this seems like an overkill. There's no reason to make Task creation any slower than it has to be; keep in mind that PyPy uses the pure Python version. Just replace this with a simple int increment.

agronholm replied:
This is what the threading module does, and it uses `itertools.count()` because `_counter += 1` is not thread safe. The same applies to asyncio if there are more than one thread running an event loop (since the counter is global).

1st1 replied:
> _counter += 1 is not thread safe.

It is thread safe in Python and will likely always be with or without the GIL.

agronholm replied:
> It is thread safe in Python and will likely always be with or without the GIL.

What output does [this script](https://gist.github.com/agronholm/c06c2dec5eb77c98ab264151fff6178c) give you then?

1st1 replied:
Hmmm... thank you for pointing this out! ;)

To my point: `+=` is actually atomic, but storing the updated number in globals is another operation and we have a race between them. This actually deserves a comment in the code explaining why we use `counter()` and not `+=`.

1st1 commented on Doc/library/asyncio-eventloop.rst:
```diff
@@ -250,7 +250,7 @@ Futures
 -----
 
-.. method:: AbstractEventLoop.create_task(coro)
+.. method:: AbstractEventLoop.create_task(coro, \*, name=None)
```
We also need to update the documentation of `set_task_factory`

agronholm replied:
Good point.

1st1 commented on Doc/library/asyncio-task.rst:
```diff
@@ -504,6 +516,12 @@ Task
       get_stack().  The file argument is an I/O stream to which the output
       is written; by default output is written to sys.stderr.
 
+   .. attribute:: name
```
I care way more about consistency of asyncio APIs than about maintaining any kind of artificial similarity between Task and Thread.  And in asyncio we don't use properties that much. Please replace the `name` property with a `Task.get_name()` method.  There's no use case for adding `set_name()` though, so let's not add it at least in 3.8.

agronholm replied:
Got it.

1st1 commented on Lib/asyncio/base_tasks.py:
```diff
@@ -12,11 +12,13 @@ def _task_repr_info(task):
 
     if task._fut_waiter is not None:
-        info.insert(2, f'wait_for={task._fut_waiter!r}')
+        info.insert(3, f'wait_for={task._fut_waiter!r}')
```
BTW, feel free to make another pull request to refactor this code. I don't like us modifying some list object with `insert()` calls.

agronholm replied:
How would you prefer that we display the name of the task in `repr()` then?

1st1 replied:
It would be great to refactor the code that formats things (tasks, callbacks, reprs) etc somehow. Right now it feels hacky and it's one of the areas of asyncio code that I'd like to be rewritten to be clearer.  Discussing that is slightly out of the scope of this PR though, but if you have any ideas how to avoid mutating lists and having a non-linear formatting logic I'd be happy to accept a new PR.  For this PR, your current code is totally fine!

1st1 commented on Lib/asyncio/tasks.py:
```diff
@@ -104,6 +108,7 @@ def __init__(self, coro, *, loop=None):
             self._log_destroy_pending = False
             raise TypeError(f"a coroutine was expected, got {coro!r}")
 
+        self._name = str(name) if name is not None else 'Task-%s' % _counter()
```
This is a bit too hard to read. I'd replace it with an `if` statement.

1st1 commented on Modules/_asynciomodule.c:
```diff
@@ -37,6 +37,8 @@ static PyObject *context_kwname;
 static volatile uint64_t cached_running_holder_tsid;
 
+/* Counter for autogenerated Task names */
+static unsigned long long task_name_counter = 0;
```
Please use `uint64_t`.

agronholm replied:
I had that at first but @ztane pointed out that the format string explicitly requires an `unsigned long long` and `uint64_t` may or may not be the same size. Are you sure this is ok?

1st1 replied:
Got it. AFAIK `unsigned long long` is guaranteed to be large enough to store `uint64_t`. Using the latter makes the code more consistent, so I'd go for it anyways.

You can use this format specifier to format `uint64_t` specifically:

```C
printf("... %" PRIu64 "...", (uint64_t)1);
```

1st1 commented on Modules/_asynciomodule.c:
```diff
@@ -1969,6 +1974,17 @@ _asyncio_Task___init___impl(TaskObj *self, PyObject *coro, PyObject *loop)
+    } else {
+        name = PyObject_Str(name);
+    }
+
```
Please remove this new line to make it easier for the reader to see that we handle the case of `name==NULL`

agronholm committed:
Made changes requested by the reviewers

* Removed the public *name* property
* Added the *Task.get_name()* method
* Modified the documentation of *set_task_factory()* to include the
  *name* parameter
* Whitespace changes

agronholm committed:
Added whatsnew entry

agronholm committed:
Changed unsigned long long to uint64_t

agronholm committed:
Added explanation for the use of itertools.count() in asyncio.tasks

1st1 reviewed (changes requested):
See https://github.com/python/cpython/pull/8547#issuecomment-409245319 for more details.

agronholm committed:
Reverted task factory API changes and added Task.set_name()

agronholm committed:
Updated the changelog entries to match the latest changes

1st1 commented on Doc/library/asyncio-task.rst:
```diff
@@ -390,7 +390,7 @@ Task
 .. function:: create_task(coro)
```
I'm confused... where's the `name` argument? Please take another look at https://github.com/python/cpython/pull/8547#issuecomment-409245319.

Perhaps the problem is that we mean different things by "task factories":

* I mean custom user factories set by `loop.set_task_factory()` (which `loop.create_task()` later calls)
* You seem to mean `loop.create_task()`.

Maybe this is where the confusion is coming from.  Please re-read https://github.com/python/cpython/pull/8547#issuecomment-409245319.

agronholm replied:
You seem to be suggesting that we ignore any `AttributeError` produced by a misbehaving task factory. This is something I've also considered and then rejected. Sweeping problems like that under the carpet is something I absolutely want to avoid. The best solution would, in my opinion, be to raise a user friendly error message explaining why setting a custom name for the task failed.

1st1 replied:
> Sweeping problems like that under the carpet is something I absolutely want to avoid. The best solution would, in my opinion, be to raise a user friendly error message explaining why setting a custom name for the task failed.

Well, since task names are not critical in *any way* for asyncio programs I tend to disagree here. They are just here to simplify debugging.

I'm OK with issuing a warning though, like "some.lib.name.Task does not have a set_name() method".  We can later (in Python 3.9 or 4?) convert this into an error.

agronholm replied:
A warning seems to be a reasonable compromise here. I will go add the `name` argument back.

1st1 replied:
Great, thanks!

agronholm replied:
For some reason I had missed parts of your commentary. It now makes a lot more sense to me. With the addition of the warning, I think we're finally on the same page. The one change I still want to make is to keep the `name` keyword argument in the stdlib Task constructor to avoid needlessly generating a task name if one was explicitly provided. Agreed?

1st1 replied:
> For some reason I had missed parts of your commentary. It now makes a lot more sense to me. With the addition of the warning, I think we're finally on the same page.

Great, I had a feeling that something is off, but couldn't pinpoint that we mean different things when we discuss "task factories" :)

> The one change I still want to make is to keep the name keyword argument in the stdlib Task constructor to avoid needlessly generating a task name if one was explicitly provided.

OK by me.

agronholm committed:
Fixed PEP 8 issue

agronholm committed:
Fixed C function name for Task.get_name()

agronholm committed:
Added back the "name" parameter to create_task() and asyncio.Task()

As suggested in the code review, this will allow us to add task names in a fashion that is backwards compatible with existing task factories and event loops.

agronholm committed:
Added back the "name" parameter to create_task()

The previous commit lacked this particular change for the bare asyncio.create_task().

agronholm committed:
Updated the news items

agronholm committed:
Added the "name" argument back to AbstractEventLoop.create_task()

agronholm committed:
Updated the news items

agronholm committed:
More documentation updates

agronholm committed:
Refactored the task name setting logic into a function

agronholm committed:
Documentation tweaks

1st1 commented on Doc/library/asyncio-eventloop.rst:
```diff
@@ -259,8 +259,14 @@ Tasks
    interoperability. In this case, the result type is a subclass of
    :class:`Task`.
 
+   If *name* is not ``None``, it is set as the name of the task using
```
I'd change to "If *name* argument is provided, ..."

1st1 commented on Modules/_asynciomodule.c:
```diff
@@ -1969,6 +1974,16 @@ _asyncio_Task___init___impl(TaskObj *self, PyObject *coro, PyObject *loop)
 
+    if (name == Py_None) {
+        name = PyUnicode_FromFormat("Task-%" PRIu64, ++task_name_counter);
+    } else {
```
Please change 

```c
} else {
``` 

to 

```c
}
else if (!PyUnicode_Check(name)) {
```

1st1 commented on Lib/asyncio/base_events.py:
```diff
@@ -393,6 +393,8 @@ def create_task(self, coro):
                 del task._source_traceback[-1]
         else:
             task = self._task_factory(self, coro)
+            tasks._set_task_name(task, name)
```
This line should be unindented or you should pass the name to `task.Task(..)` few lines above.  Please add a test that Tasks get custom names when created via `loop.create_task(name=..)` with a standard factory.

agronholm committed:
Made the changes requested by the reviewer

agronholm committed:
Fixed segfault if "name" is already a string

agronholm committed:
Fixed the new test to use a custom factory

1st1 commented on Lib/asyncio/tasks.py:
```diff
@@ -104,6 +120,11 @@ def __init__(self, coro, *, loop=None):
             raise TypeError(f"a coroutine was expected, got {coro!r}")
 
+        if name is None:
+            self._name = 'Task-%s' % _task_name_counter()
```
nit: I've recently updated asyncio to use f-strings. Please change this one to either an f-string or to a `.format()` call (just for maintaining consistency of the codebase).

1st1 commented on Lib/test/test_asyncio/test_base_events.py:
```diff
@@ -813,6 +813,22 @@ def create_task(self, coro):
         task._log_destroy_pending = False
         coro.close()
 
+    def test_create_named_task_with_custom_factory(self):
```
Please also add `test_create_named_task_with_default_factory`. That would test a bug we fixed in this PR when the default task factory didn't pass the `name=name` to `Task()`.

agronholm committed:
Fixed last two nits

1st1 commented on Modules/_asynciomodule.c:
```diff
@@ -2297,6 +2316,41 @@ _asyncio_Task_set_exception(TaskObj *self, PyObject *exception)
+{
+    PyObject *name = PyObject_Str(value);
+    if (name == NULL) {
+        return -1;
```
Oups, need to fix this to `return NULL`!

agronholm replied:
With `make clean;  CFLAGS=-Wno-cast-function-type make`, compilation of the `_asyncio` extension produced no warnings after the patch.

1st1 reviewed (changes requested):
Just spotted an error, a `PyObject*` function returns `-1`.  Please also check that the C compiler generates 0 warnings for your patch in `_asynciomodule.c`.

agronholm committed:
Fixed incompatible return value from set_name()
